log/scope.go

```diff
@@ -315,3 +347,103 @@ func (s *Scope) SetLogCallers(logCallers bool) {
 func (s *Scope) GetLogCallers() bool {
 	return s.logCallers.Load().(bool)
 }
+
+// Copy makes a copy of s and returns a pointer to it.
+func (s *Scope) Copy() *Scope {
+	out := *s
+	out.labels = copyStringInterfaceMap(s.labels)
+	return &out
+}
+
+// WithLabels adds a key-value pairs to the labels in s. The key must be a string, while the value may be any type.
+// It returns a copy of s, with the labels added.
+// e.g. newScope := oldScope.WithLabels("foo", "bar", "baz", 123, "qux", 0.123)
+func (s *Scope) WithLabels(kvlist ...interface{}) *Scope {
+	out := s.Copy()
+	if len(kvlist)%2 != 0 {
+		out.labels["WithLabels error"] = fmt.Sprintf("even number of parameters required, got %d", len(kvlist))
+		return out
+	}
+
+	for i := 0; i < len(kvlist); i += 2 {
+		keyi := kvlist[i]
+		key, ok := keyi.(string)
+		if !ok {
+			out.labels["WithLabels error"] = fmt.Sprintf("label name %v must be a string, got %T ", keyi, keyi)
+			return out
+		}
+		out.labels[key] = kvlist[i+1]
+		out.labelKeys = append(out.labelKeys, key)
+	}
+	return out
+}
+
+// WithoutLabels makes a copy of s, clears labels in s with the given keys and returns the copy.
+// Not-existent keys are ignored.
+// e.g. myScope = myScope.WithLabels("foo", "bar", "baz", 123, "qux", 0.123)
+//      myScope = myScope.WithoutLabels("baz", "qux")
+// Result: myScope retains only the "foo" label.
+func (s *Scope) WithoutLabels(keys ...string) *Scope {
```

Comment:
nit: do we need this? I would expect usage to be

```
s := some-scope
if ... {
  s = s.WithLabels(foo)
  s.Log(something with foo)
}
s.Log(something without foo)
```

Reply:
Basically I am biased in favor of minimal API until proven a need otherwise. We can always add, hard to remove and use a big API though. But no strong opinion here

Reply:
that's a good question. i guess i took it for granted that someone would want to do this but when pressed i'm having a hard time coming up with a convincing example. i will remove for now, we can indeed easily add it back.